src/Ocelot.Provider.Kubernetes/Kube.cs

```diff
@@ -26,28 +30,39 @@ public class Kube : IServiceDiscoveryProvider
         _logger = factory.CreateLogger<Kube>();
         _kubeApi = kubeApi;
         _serviceBuilder = serviceBuilder;
-        _services = new();
     }
 
     public virtual async Task<List<Service>> GetAsync()
     {
-        var endpoint = await _kubeApi
-            .ResourceClient(client => new EndPointClientV1(client))
-            .GetAsync(_configuration.KeyOfServiceInK8s, _configuration.KubeNamespace);
+        var endpoint = await Retry.OperationAsync(GetEndpoint, Ensure, logger: _logger);
 
-        _services.Clear();
-        if (endpoint?.Subsets.Count != 0)
+        if (Ensure(endpoint))
         {
-            _services.AddRange(BuildServices(_configuration, endpoint));
+            _logger.LogWarning(() => GetMessage($"Unable to use bad result returned by {nameof(Kube)} integration endpoint because the final result is invalid/unknown after multiple retries!"));
+            return new(0);
         }
-        else
+
+        return BuildServices(_configuration, endpoint)
+            .ToList();
+    }
+
+    private Task<EndpointsV1> GetEndpoint() => _kubeApi
+        .ResourceClient(client => new EndPointClientV1(client))
+        .GetAsync(_configuration.KeyOfServiceInK8s, _configuration.KubeNamespace);
+
+    private bool Ensure(EndpointsV1 endpoint)
```

Comment:
I'm not sure about this method. Do we really need it? First, we are writing two warning messages here, the first one in ```Ensure``` the second one in ```GetAsync```.  I don't like methods returning true or false explicitely...

Reply:
> I don't like methods returning true or false explicitly...

Fair enough! I don't like seeing old Tom's code with the same code bool-pattern 🤣 
Fixed via warning logging removal.

Reply:
Your suggestions were followed, see changes https://github.com/ThreeMammals/Ocelot/pull/2111/commits/c4c7ae0ff2a891d64b4f81b6a6d051cc15e71211#diff-9b178d28d53bc91aa0d4b24da98f78648855928a824eabde6eb3a87acb115763